### ISSUE TITLE
Saving the state of the model when exiting the program?

### ISSUE THREAD

**Issue by Vladonai** [LostRuins/koboldcpp#1240]
Is there now a possibility to add this feature to the program: when exiting the program (or pressing CRTL+C, as I usually do), if the -savestate key was used at startup, then save the state of the model, so that the next startup does not recalculate the KV-cache, but just load it from disk?

I have 4 Tesla P40's, using the 123B_4KM model and a 24k context. I really need this feature!

**Comment by jojorne** [comment]
Every user that uses the same server ends up killing the cache, right? It is recalculated from scratch for each request because the context changes from person to person. Keeping the cache of all users in memory would end up using a lot of memory. But, having a computer with little memory, I understand the desire to save the cache. I can only imagine the time it takes with this model.

**Comment by LostRuins** [comment]
Yes, the cache will be rewound back to the first diverging token. But normally prompt processing is fast enough that you won't realize it. Jojorne is right that multiple cache saving is not really feasible for a server application.

**Comment by Vladonai** [comment]
Kobold is used as backend, Silly Tavern as frontend. Model 123B_Q4KM, 24k contexts. In the last session, the context in Tavern was completely filled in. Let's start a new session and pick up where we left off:
![24k](https://github.com/user-attachments/assets/cc4fcb07-d94f-44ea-9a30-714c82c6a7fa)
If instead of recalculating the entire context over again, the context cache were loaded along with the model, this time would be significantly reduced.

**Comment by LostRuins** [comment]
The kv cache for 123B will be many gigabytes too though.

**Comment by Vladonai** [comment]
> The kv cache for 123B will be many gigabytes too though.

I understand that, and I accept it.

**Comment by MrReplikant** [comment]
Just making it known here, I too would like to see this implemented, if for no other reason than to overcome the long startup times like the one pictured above

**Comment by LostRuins** [comment]
Hi all, this will finally be added. KV save and load will be available over the admin api, you simply call one function to SaveState and another function to LoadState. States will be saved in RAM, and reusable multiple times by reloading it. Context will automatically be restored to the saved state when it's reloaded, allowing you to avoid reprocessing.

https://github.com/LostRuins/koboldcpp/commit/4b57108508e80b49556e3c33670c923c9d3818c2

**Comment by LostRuins** [comment]
Now available
